daemon/api.go

```diff
@@ -157,6 +164,59 @@ func sysInfo(c *Command, r *http.Request) Response {
 	return SyncResponse(m)
 }
 
+type authState struct {
+	Users []userAuthState `json:"users"`
+}
+
+type userAuthState struct {
+	Username   string   `json:"username,omitempty"`
+	Macaroon   string   `json:"macaroon,omitempty"`
+	Discharges []string `json:"discharges,omitempty"`
+}
+
+func loginUser(c *Command, r *http.Request) Response {
+	var loginData struct {
+		Username string `json:"username"`
+		Password string `json:"password"`
+		Otp      string `json:"otp"`
+	}
+
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&loginData); err != nil {
+		return BadRequest("cannot decode login data from request body: %v", err)
+	}
+
+	macaroon, err := store.RequestPackageAccessMacaroon()
+	if err != nil {
+		return InternalError("cannot get package access macaroon")
+	}
+
+	discharge, err := store.DischargeAuthCaveat(loginData.Username, loginData.Password, macaroon, loginData.Otp)
+	if err == store.ErrAuthenticationNeeds2fa {
+		twofactorRequiredResponse := &resp{
+			Type: ResponseTypeError,
+			Result: &errorResult{
+				Kind:    errorKindTwoFactorRequired,
+				Message: "two factor authentication required"},
+			Status: http.StatusUnauthorized}
+		return SyncResponse(twofactorRequiredResponse)
+	}
+	if err != nil {
+		return Unauthorized("cannot get discharge authorization")
+	}
+
+	authenticatedUser := userAuthState{Username: loginData.Username, Macaroon: macaroon, Discharges: []string{discharge}}
+	authStateData := authState{Users: []userAuthState{authenticatedUser}}
+
+	overlord := c.d.overlord
+	state := overlord.State()
+	state.Lock()
+	state.Set("auth", authStateData)
+	state.Unlock()
+
+	return SyncResponse(authenticatedUser)
```

Comment:
Please cook a local type here, containing just the macaroon and the discharges, so we untie the internal state representation from the external API result. We won't send the username they sent us back, for example.